### PR TITLE
objectives ecs rework

### PR DESCRIPTION
**Pull request by deltanedas**
## About the PR
except recommitted

- **objectives are entities now**
- removed support for multiple conditions on 1 objective since nothing used it and it made the code more complex for no reason
- split up condition logic to be more customizable both from other condition components and in yaml
- renamed KillRandomObjective to KillRandomPersonObjective
- fixed order bug where if you first roll help traitor progress objective, you can then get kill objectives after that. but it didnt work the other way around. now you cant get social and kill objectives ever
- removed a condition data duplication check since it isnt (to my knowledge) possible to reimplement cleanly with components and if you get 2 kill objectives for the same guy hes very unlucky and you REALLY need that person dead

gameplay changes:
- increased nuke disk difficulty so you dont get stretched out too thin with stealing disk+jordans and escaping on the shuttle which is basically a sosa exclusive to complete
- same for secret documents, hos has wt550 roundstart ce does not, and all cap objectives besides disk
- changed disk requirements to check if you are command, previously it allowed qm for some reason
  personally i think anyone that isnt captain should still be able to get it, but i just exclude qm too for consistency with the rest of the requirements
- finally grouped objectives from multiple game rules of the same type

## Why / Balance
the gameplay changes have reasons next to them :trollface:

## Technical details
at its core ObjectiveComponent must be on an objective entity
after that add requirements a condition and other stuff then boom a task
the yaml uses ok prototype composition and should be fairly reusable

4 main events added:
- `ObjectiveGetProgressEvent` - gets an objectives progress. condition components have to handle it and set progress or it wont be doable.
- `ObjectiveAssignedEvent` - when an objective is assigned components can set state like randomly picking a kill target. the new version of IObjectiveCondition
- `ObjectiveAfterAssignEvent` - raised after everything has handled assigned and hasnt been cancelled, can set objective name description and icon from a system here.
- `RequirementCheckEvent` - raised before assigning an objective, makes sure theres no hard conflicts with other objectives or the player etc. the new version of IObjectiveRequirement

## Media
tasks exist

yep got them jordans

kill works too!!!

assassination too

ninja objectives updated

multiple gamerules (each ninja spawns its own rule) are grouped together

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
- `Objective`, `ObjectivePrototype` and objective condition/requirement interfaces have been removed, objectives are now entities with components.
- multiple objective conditions were unused so they are no longer supported. just use multiple objectives
- `SharedMindSystem.TryAddObjective` now takes an entity prototype id instead of a now removed objective prototype.
- `ConditionInfo` is now `ObjectiveInfo`
- all objective prototypes have been turned into entities which use various parents in base_objectives.yml, use them for any new objectives
- for translators, objectives titles and descriptions that don't use any data like targets or counts are translated like other entities. Anything formatted will still use objectives/conditions/whatever.ftl

**Changelog**
:cl:
- tweak: Some Syndicate objectives have been tweaked to have higher difficulties, notably stealing the nuke disk.

### CODE REVIEW

**Comment by deltanedas** [comment]
should be somewhat easier to review, recommitted everything

**Comment by deltanedas** [comment]
merg...

**Comment by deltanedas** [comment]
@metalgearsloth pls merge i need this so bad

File: Content.Shared/Objectives/Systems/ObjectiveSystem.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+        var info = ev.Info;
+        if (info.Title == null || info.Description == null || info.Icon == null || info.Progress == null)
+        {
+            Log.Error($"An objective {uid} of {_mind.MindOwnerLoggingString(mind)} has incomplete info: {info.Title} {info.Description} {info.Progress}");
+            info.Title ??= "!!!BROKEN OBJECTIVE!!!";
+            info.Description ??= "!!! BROKEN OBJECTIVE DESCRIPTION!!!";
+            info.Icon ??= new SpriteSpecifier.Rsi(new ("error.rsi"), "error");
+            info.Progress ??= 0f;
+        }
```
Comment (DrSmugleaf) [inline Content.Shared/Objectives/Systems/ObjectiveSystem.cs:123]: This shouldn't return an ObjectiveInfo instance if the info creation fails. Also are info description and icon strictly necessary for an objective info to exist and be displayed?

File: Content.Shared/Objectives/Systems/ObjectiveSystem.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+using Content.Shared.Objectives.Components;
+using Robust.Shared.Utility;
+
+namespace Content.Shared.Objectives.Systems;
+
+/// <summary>
+/// Provides API for creating and interacting with objectives.
+/// Adds default info from <see cref="ObjectiveComponent"/> to <see cref="ObjectiveGetInfoEvent"/>.
+/// </summary>
+public sealed class ObjectiveSystem : EntitySystem
```
Comment (DrSmugleaf) [inline Content.Shared/Objectives/Systems/ObjectiveSystem.cs:12]: This should really be merged with ObjectivesSystem, or just any option other than having both ObjectiveSystem and Objective**s**System.

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

File: Content.Shared/Objectives/Systems/ObjectiveSystem.cs (hunk 1 above)
Comment (deltanedas) [reply Content.Shared/Objectives/Systems/ObjectiveSystem.cs:123]: not necessarily but even the lowest quality objective should have an icon

File: Content.Shared/Objectives/Systems/ObjectiveSystem.cs (hunk 2 above)
Comment (deltanedas) [reply Content.Shared/Objectives/Systems/ObjectiveSystem.cs:12]: yes it is unfortunate

File: Content.Shared/Objectives/Components/ObjectiveComponent.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    /// </summary>
+    [DataField("title"), ViewVariables(VVAccess.ReadWrite)]
+    public string? Title;
+
+    /// <summary>
+    /// Optional locale id to handle <see cref="ObjectiveGetInfoEvent"/>'s description.
+    /// If another component provides it do not use this.
+    /// </summary>
+    [DataField("description"), ViewVariables(VVAccess.ReadWrite)]
+    public string? Description;
```
[the remark this answers is not in the document]
Comment (deltanedas) [reply Content.Shared/Objectives/Components/ObjectiveComponent.cs:47]: for normal stuff would it be like

```
- type: entity
  parent: [BaseNinjaObjective, BaseSurviveObjective]
  id: NinjaSurviveObjective
  name: Survive
  description: You wouldn't be a very good ninja if you died.
```

and then if someone wants to translate it they do the ent.SurviveObjective thing

**Comment by Slava0135** [comment]
fix your shit
`space-station-14/Content.Server/Objectives/Systems/HelpProgressConditionSystem.cs(18,35): error CS0246: The type or namespace name 'ObjectiveSystem' could not be found (are you missing a using directive or an assembly reference?)`

**Comment by Equivocateur** [comment]
Noob didn't use refactor feature in Rider

File: Content.Shared/Objectives/Components/ObjectiveComponent.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mirrorcult) [reply Content.Shared/Objectives/Components/ObjectiveComponent.cs:47]: yes

**Comment by deltanedas** [comment]
still works :trollface:

**Review by DrSmugleaf — APPROVED** [review]

**Pull request merged**